src/ast-writer.cc

```diff
@@ -58,302 +58,338 @@ enum class NextChar {
+  void WriteOpenSpace(const char* name);
+  void WriteClose(NextChar next_char);
+  void WriteCloseNewline();
+  void WriteCloseSpace();
+  void WriteString(const std::string& str, NextChar next_char);
+  void WriteStringSlice(const StringSlice* str, NextChar next_char);
+  bool WriteStringSliceOpt(const StringSlice* str, NextChar next_char);
+  void WriteStringSliceOrIndex(const StringSlice* str,
+                               uint32_t index,
+                               NextChar next_char);
+  void WriteQuotedData(const void* data, size_t length);
+  void WriteQuotedStringSlice(const StringSlice* str, NextChar next_char);
+  void WriteVar(const Var* var, NextChar next_char);
+  void WriteBrVar(const Var* var, NextChar next_char);
+  void WriteType(Type type, NextChar next_char);
+  void WriteTypes(const TypeVector& types, const char* name);
+  void WriteFuncSigSpace(const FuncSignature* func_sig);
+  void WriteBeginBlock(const Block* block, const char* text);
+  void WriteEndBlock();
+  void WriteBlock(const Block* block, const char* start_text);
+  void WriteConst(const Const* const_);
+  void WriteExpr(const Expr* expr);
+  void WriteExprList(const Expr* first);
+  void WriteInitExpr(const Expr* expr);
+  void WriteTypeBindings(const char* prefix,
+                         const Func* func,
+                         const TypeVector& types,
+                         const BindingHash& bindings);
+  void WriteFunc(const Module* module, const Func* func);
+  void WriteBeginGlobal(const Global* global);
+  void WriteGlobal(const Global* global);
+  void WriteLimits(const Limits* limits);
+  void WriteTable(const Table* table);
+  void WriteElemSegment(const ElemSegment* segment);
+  void WriteMemory(const Memory* memory);
+  void WriteDataSegment(const DataSegment* segment);
+  void WriteImport(const Import* import);
+  void WriteExport(const Export* export_);
+  void WriteFuncType(const FuncType* func_type);
+  void WriteStartFunction(const Var* start);
+
+  Stream stream_;
+  Result result_ = Result::Ok;
+  int indent_ = 0;
+  NextChar next_char_ = NextChar::None;
+  int depth_ = 0;
+  std::vector<std::string> index_to_name_;
+
+  int func_index_ = 0;
+  int global_index_ = 0;
+  int table_index_ = 0;
+  int memory_index_ = 0;
+  int func_type_index_ = 0;
 };
 
 }  // namespace
 
-static void indent(Context* ctx) {
-  ctx->indent += INDENT_SIZE;
+void ASTWriter::Indent() {
+  indent_ += INDENT_SIZE;
 }
 
-static void dedent(Context* ctx) {
-  ctx->indent -= INDENT_SIZE;
-  assert(ctx->indent >= 0);
+void ASTWriter::Dedent() {
+  indent_ -= INDENT_SIZE;
+  assert(indent_ >= 0);
 }
 
-static void write_indent(Context* ctx) {
+void ASTWriter::WriteIndent() {
   static char s_indent[] =
       "                                                                       "
       "                                                                       ";
   static size_t s_indent_len = sizeof(s_indent) - 1;
-  size_t indent = ctx->indent;
```

Comment:
Why did you move from a local variable to the field? Doesn't this change the effect (you change the indent of the writer)?